libxl: set script field of libxl_device_disk

Add a hack to the libvirt libxl driver to set
libxl_device_disk->script when the disk configuration starts
with some well-known Xen external block scripts: dmmd, drbd,
and npiv.

For more details, see bsc#954872 and FATE#319810

Index: libvirt-6.2.0/src/libxl/libxl_conf.c
===================================================================
--- libvirt-6.2.0.orig/src/libxl/libxl_conf.c
+++ libvirt-6.2.0/src/libxl/libxl_conf.c
@@ -870,6 +870,22 @@ libxlDiskSetDiscard(libxl_device_disk *x
 #endif
 }
 
+static int
+libxlDiskSetScript(libxl_device_disk *x_disk, const char *disk_spec)
+{
+    if (disk_spec == NULL)
+        return 0;
+
+    if (STRPREFIX(disk_spec, "dmmd:"))
+        x_disk->script = g_strdup("block-dmmd");
+    else if (STRPREFIX(disk_spec, "drbd:"))
+        x_disk->script = g_strdup("block-drbd");
+    else if (STRPREFIX(disk_spec, "npiv:"))
+        x_disk->script = g_strdup("block-npiv");
+
+    return 0;
+}
+
 static void
 libxlDiskSetCacheMode(libxl_device_disk *x_disk, int cachemode)
 {
@@ -1010,6 +1026,7 @@ libxlMakeNetworkDiskSrc(virStorageSource
 int
 libxlMakeDisk(virDomainDiskDefPtr l_disk, libxl_device_disk *x_disk)
 {
+    const char *src = virDomainDiskGetSource(l_disk);
     const char *driver = virDomainDiskGetDriver(l_disk);
     int format = virDomainDiskGetFormat(l_disk);
     int actual_type = virStorageSourceGetActualType(l_disk->src);
@@ -1025,7 +1042,7 @@ libxlMakeDisk(virDomainDiskDefPtr l_disk
         if (libxlMakeNetworkDiskSrc(l_disk->src, &x_disk->pdev_path) < 0)
             return -1;
     } else {
-        x_disk->pdev_path = g_strdup(virDomainDiskGetSource(l_disk));
+        x_disk->pdev_path = g_strdup(src);
     }
 
     x_disk->vdev = g_strdup(l_disk->dst);
@@ -1136,6 +1153,9 @@ libxlMakeDisk(virDomainDiskDefPtr l_disk
     if (libxlDiskSetDiscard(x_disk, l_disk->discard) < 0)
         return -1;
     libxlDiskSetCacheMode(x_disk, l_disk->cachemode);
+    if (libxlDiskSetScript(x_disk, src) < 0)
+        return -1;
+
     /* An empty CDROM must have the empty format, otherwise libxl fails. */
     if (x_disk->is_cdrom && !x_disk->pdev_path)
         x_disk->format = LIBXL_DISK_FORMAT_EMPTY;
